src/lib/optimizer/strategy/data_induced_predicate_rule.cpp

```diff
@@ -0,0 +1,130 @@
+#include "data_induced_predicate_rule.hpp"
+
+#include "cost_estimation/abstract_cost_estimator.hpp"
+#include "expression/binary_predicate_expression.hpp"
+#include "expression/expression_functional.hpp"
+#include "expression/expression_utils.hpp"
+#include "logical_query_plan/aggregate_node.hpp"
+#include "logical_query_plan/enable_make_for_lqp_node.hpp"
+#include "logical_query_plan/join_node.hpp"
+#include "logical_query_plan/lqp_utils.hpp"
+#include "logical_query_plan/predicate_node.hpp"
+#include "logical_query_plan/projection_node.hpp"
+#include "statistics/abstract_cardinality_estimator.hpp"
+
+namespace hyrise {
+using namespace expression_functional;  // NOLINT(build/namespaces)
+
+std::string DataInducedPredicateRule::name() const {
+  static const auto name = std::string{"DataInducedParameterRule"};
+  return name;
+}
+
+void DataInducedPredicateRule::_apply_to_plan_without_subqueries(
+    const std::shared_ptr<AbstractLQPNode>& lqp_root) const {
+  Assert(lqp_root->type == LQPNodeType::Root, "Rule needs root to hold onto");
+
+  // Adding dips inside visit_lqp might lead to endless recursions. Thus, we use visit_lqp to identify the
+  // reductions that we want to add to the plan, write them into data_induced_predicates and actually add them after
+  // visit_lqp.
+  std::vector<std::tuple<std::shared_ptr<JoinNode>, LQPInputSide, std::shared_ptr<PredicateNode>>>
+      data_induced_predicates;
+
+  const auto opposite_side = [](const auto side) {
+    return side == LQPInputSide::Left ? LQPInputSide::Right : LQPInputSide::Left;
+  };
+
+  const auto estimator = cost_estimator->cardinality_estimator->new_instance();
+  estimator->guarantee_bottom_up_construction();
+
+  visit_lqp(lqp_root, [&estimator, &opposite_side, &data_induced_predicates](const auto& node) {
+    if (node->type != LQPNodeType::Join) {
+      return LQPVisitation::VisitInputs;
+    }
+    const auto join_node = std::static_pointer_cast<JoinNode>(node);
+
+    for (const auto& join_predicate : join_node->join_predicates()) {
+      const auto predicate_expression = std::dynamic_pointer_cast<BinaryPredicateExpression>(join_predicate);
+      DebugAssert(predicate_expression, "Expected BinaryPredicateExpression");
```

Comment:
We oriented our code at the SemiJoinReductionRule which also used DebugAsserts. We assumed that this was the right amount of escalation in case of unexpected inputs.

Reply:
Because this is an optimizer rule, the build would be just fine, we just couldn't optimize in this case. So I think a DebugAssert, as Sandro mentioned, is sufficient enough here

Reply:
Ok, that makes sense. Thanks for explaining!